src/cmd/cmdmap.cpp

```diff
@@ -1,42 +1,131 @@
 #include "cmdmap.h"
 #include "logger.h"
+#include "cmdimp.h"
 
 std::string CommandMap::get_usage(std::string command) {
   // Check if command exists?
+  std::string result;
   try {
-    return info_map[command].usage;
+    result = info_map[command]->usage;
   }
   catch(int e) {
     Logger::info << "get_usage(" << command << ") throws error" << e;
-    exit(1);
+    result = "Error: usage not found";
   }
+  return result;
 }
 
 int CommandMap::exec_command(cmdargs &cmd, std::string command) {
+  // Check if the commands vector contains command
+  bool seen = false;
+  for(unsigned int i = 0; i < commands.size(); i++) {
+    if(command == commands[i])
+      seen = true;
+  }
+  if(!seen) {
+    Logger::fatal << "Unrecognized command: " << command << Logger::endl;
+    return -1;
+  }
+
+  int result;
   try {
-    return info_map[command].helper_fn(cmd);
+    result =  info_map[command]->helper_fn(cmd);
   }
   catch(int e) {
     Logger::info << "exec_command(" << command << ") throws error" << e;
-    exit(1);
+    result = -1;
   }
+  return result;
 }
 
-int CommandMap::add_command(std::string cmd_name, std::string cmd_usage, int (* helper_fn)(cmdargs cmd)) {
-  // Add the command to the vector of commands
-  commands.push_back(cmd_name);
+CommandMap init_autolab_command_map() {
 
-  // Construct the command_info struct for this command
-  command_info cmd_info;
-  cmd_info.usage = cmd_usage;
-  cmd_info.helper_fn = helper_fn;
-  info_map[cmd_name] = cmd_info;
+  std::vector<std::string> valid_commands = {"status", "download", "submit",
+                                             "courses", "assessments", "asmts",
+                                             "problems", "scores", "feedback",
+                                             "enroll"};
 
-  // Why am I returning an int again?
-  return 0;
-}
+  // There's gotta be a better way to do this
```

Comment:
did you mean to come up with a better way? :)

Reply:
Oh, haha. Yeah, that was a complete mess back when I was trying to use malloc() to instantiate the map. I guess we can delete the comment now :P

Reply:
haha alright